src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,289 @@
+        operator Xbyak::RegExp() const { ensureValid(); return reg; }
+        int getIdx() const { ensureValid(); return reg.getIdx(); }
+        friend Xbyak::RegExp operator+(const Reg& lhs, const Xbyak::RegExp& rhs) {
+            lhs.ensureValid();
+            return lhs.operator Xbyak::RegExp() + rhs;
+        }
+        void release() {
+            if (regPool) {
+                regPool->returnToPool(reg);
+                regPool.reset();
+            }
+        }
+
+    private:
+        void ensureValid() const {
+            if (!regPool) {
+                IE_THROW() << "RegistersPool::Reg is either not initialized or released";
+            }
+        }
+
+        void initialize(const RegistersPool::Ptr& pool, int fixedIdx = anyIdx) {
+            release();
+            reg = TReg(pool->template getFree<TReg>(fixedIdx));
+            regPool = pool;
+        }
+
+    private:
+        TReg reg;
+        RegistersPool::Ptr regPool;
+    };
+
+    virtual ~RegistersPool() {
+        checkUniqueAndUpdate(false);
+    }
+
+    template <x64::cpu_isa_t isa>
+    static Ptr create(std::initializer_list<Xbyak::Reg> regsToExclude);
+
+    template<typename TReg>
+    size_t countFree() const {
+        if (std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value) {
+            return simdSet.countUnused();
+        } else if (std::is_same<TReg, Xbyak::Reg64>::value || std::is_same<TReg, Xbyak::Reg32>::value) {
+            return generalSet.countUnused();
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            return countUnusedOpmask();
+        }
+    }
+
+protected:
+    class PhysicalSet {
+        static constexpr int undefinedIdx = -1;
+    public:
+        PhysicalSet(int size) : isFreeIndexVector(size, true) {}
+
+        void setAsUsed(int regIdx) {
+            if (regIdx >= isFreeIndexVector.size() || regIdx < 0) {
+                IE_THROW() << "regIdx is out of bounds in RegistersPool::PhysicalSet::setAsUsed()";
+            }
+            if (!isFreeIndexVector[regIdx]) {
+                IE_THROW() << "Inconsistency in RegistersPool::PhysicalSet::setAsUsed()";
+            }
+            isFreeIndexVector[regIdx] = false;
+        }
+
+        void setAsUnused(int regIdx) {
+            if (regIdx >= isFreeIndexVector.size() || regIdx < 0) {
+                IE_THROW() << "regIdx is out of bounds in RegistersPool::PhysicalSet::setAsUsed()";
+            }
+            if (isFreeIndexVector[regIdx]) {
+                IE_THROW() << "Inconsistency in RegistersPool::PhysicalSet::setAsUnused()";
+            }
+            isFreeIndexVector[regIdx] = true;
+        }
+
+        int getUnused(int fixedIdx) {
+            if (fixedIdx == anyIdx) {
+                auto idx = getFirstFreeIndex();
+                if (idx == undefinedIdx) {
+                    IE_THROW() << "Not enough registers in the RegistersPool";
```

Comment:
Why not throw this exception in `getFirstFreeIndex()` ?

Reply:
Also it looks like `getFirstFreeIndex` is used only in `getUnused`, it means that body of `getFirstFreeIndex` could be moved in `getUnused`

Reply:
Throwing the exception from the `getFirstFreeIndex()` is a good point. But I think it is good to have a separate method `getFirstFreeIndex()` because (1) it is suitable for readability and (2) I get rid of the `undefinedIdx`.

Reply:
Okay, lets leave it like this now